### ISSUE TITLE
Name and other methods

### ISSUE THREAD

**Issue by trowski** [thenable/Thenable#2]
There are two major points regarding this interface that I believe need to be discussed:

1. Should the name of the interface be `Thenable` or `ThenableInterface`?
2. Should the interface include a method `otherwise()` that defines a callback that is invoked if the thenable is rejected. This method could use reflection to examine type-hints and invoke the registered callback only if the type-hints match or could simply be an alias of `then(null, $onRejected)` like JavaScript's `Promise.catch()` method.

**Comment by trowski** [comment]
Potential third issue to discuss: Adding a `done()` method that does not return another thenable. Registered callbacks are executed upon resolution, ignoring return values and throwing exceptions in an uncatchable way.

This method has been implemented by many popular JS promise libraries.

**Comment by coderstephen** [comment]
For the name, I definitely recommend `Thenable` - it's catchy, short, and I tire of `Interface` suffixes to interfaces that have good interface-sounding names already. My 2¢.

**Comment by kelunik** [comment]
`Thenable`, for sure. Adding suffixes is a bad habit.

**Comment by trowski** [comment]
Since opening this issue, I realized that PHP-FIG requires all interfaces to have a `-Interface` suffix. I think `ThenableInterface` sounds odd, so I changed the name to `PromiseInterface`. The name is still open for discussion, but I felt `PromiseInterface` was more immediately recognizable than `ThenableInterface`.

**Comment by kelunik** [comment]
I know that PHP-FIG requires it. We don't need a promise specification to be a PSR...

**Comment by sagikazarmark** [comment]
> We don't need a promise specification to be a PSR

Actually I would be happy to see it as a PSR.

**Comment by kelunik** [comment]
> Actually I would be happy to see it as a PSR.

Sure, but I just meant that we could also have a common standard for react, icicle and amp without the suffix if it's not a PSR. It might even be good to separate things, since mixing synchronous and asynchronous interfaces isn't a good idea anyway.

We could have an async alliance without that stupid suffix requirement. ;-P

**Comment by sagikazarmark** [comment]
> We could have an async alliance without that stupid suffix requirement. ;-P

Totally agree, I got rid of Interface suffixes years ago.

**Comment by trowski** [comment]
I don't like the Interface suffix either and will be switching the name back to `Thenable`. In fact, I'm dropping them from Icicle completely. I only used them to begin with since it seemed common in PHP, but you can't change horrible patterns if you keep repeating them. :disappointed:

**Comment by sagikazarmark** [comment]
http://verraes.net/2013/09/sensible-interfaces/ - by @mathiasverraes

I read it every morning and evening.

**Comment by trowski** [comment]
Yep, fantastic article – agree with everything he says. I just got so used to having the Interface suffix on everything it took some adjustment, but now I greatly prefer it.

**Comment by sagikazarmark** [comment]
Yeah, habits...

But when you start to **think** (tm) then you see how bad it is.

**Comment by jsor** [comment]
1. I'm undecided about the `Interface` suffix. I'm also not using it in private/work code but i'm using it in my open-source libraries just because it is so common. I'd be fine with both.
2. There are reasons for and against adding `otherwise`/`except`. I've [commented here](https://github.com/async-interop/promise/commit/657238a4ec23e23984565416db4595171cbe0955#commitcomment-13701747) about the pro's. If we keep it, i'd vote for removing the second argument from `then`, though this would differ from how it is commonly implemented, especially in the JS world. A reason against adding it would be, that the behavior (support of type-hint based execution) isn't clearly defined and consumers cannot really program against it. All in all, i'm now more against adding it.
3. :+1: for adding `done`.

**Comment by sagikazarmark** [comment]
Just checked the Promises/A+ spec: Thenable is a perfectly valid name for an object having the then method. So I would say having a Thenable with then only, and a Promise with other stuff could probably work.

**Comment by trowski** [comment]
I had planned on changing the interface name back to `Thenable`, but then this repo was moved into this group with a goal of creating PSRs. I would still love to drop the `-Interface` suffix and choose a more generic name like `Thenable` or `Awaitable`. We could drop the suffix within this group, but the PSR versions will require the suffix (unless we can convince them otherwise… unlikely).

Regarding `otherwise()` or `except()` methods, I've come to a similar conclusion as @jsor. There are reasons for and against having the methods, with behavior of parameter type-hints being one of the biggest reasons to avoid adding it to the interface. I think it would be best to not include an error-only method, instead keeping the interface simple with a JS-like `then()` method returning another promise plus a void `done()` method for consuming the result. Two methods only, that's all that's really needed in a basic interface. Libraries can provide wrappers to morph objects implementing our basic interface into whatever promise interface they provide.

**Comment by sagikazarmark** [comment]
I can live with the Interface suffix, it's a necessary evil to create a PSR.
I would still try to create multiple interfaces (like there are multiple specs for A+):

- Thenable: have then (and done)
- Awaitable: have wait
- Promise: extend thenable (and have done and other stuff here)?

This way implementations can be separated from a functional point of view, but still can be used for example simply passing around a promise and calling then. In those cases, it doesn't matter if you can wait for it.

**Comment by jeskew** [comment]
Would it be possible to have the Thenable interface mentioned include only `then`? This is often the only method promise consumers will interact with directly, whereas `done` seems like it could be covered in the extended promise interface described.

Also, would it be out of scope for this repository to include some version of a Deferred object? For example, Guzzle's [`PromisorInterface`](https://github.com/guzzle/promises/blob/master/src/PromisorInterface.php).
